feat(delete range): optimize delete range

soundOfDestiny opened the pull request:
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

<!--

**This section will be used as the commit message. Please do not leave this empty!**

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Refer to a related PR or issue link (optional)

-->
optimize delete range to solve #8872

Previously delete range might be cloned to many ranges before saved into SST, in this PR we can save the original delete ranges.

Assume that watermark1 is 5, watermark2 is 7, watermark3 is 11, delete ranges
`{ [0, wmk1) in epoch1, [wmk1, wmk2) in epoch2, [wmk2, wmk3) in epoch3 }`
can be transformed into events below:
`{ <0, +epoch1> <wmk1, -epoch1> <wmk1, +epoch2> <wmk2, -epoch2> <wmk2, +epoch3> <wmk3,
-epoch3> }`
Then we can get monotonic events (they are in order by user key) as below:
`{ <0, epoch1>, <wmk1, epoch2>, <wmk2, epoch3>, <wmk3, +inf> }`
which means that delete range of [0, wmk1) is epoch1, delete range of [wmk1, wmk2) if
epoch2, etc. In this example, at the event key wmk1 (5), delete range changes from
epoch1 to epoch2, thus the `new epoch` is epoch2. epoch2 will be used from the event
key wmk1 (5) and till the next event key wmk2 (7) (not inclusive).
If there is no range deletes between current event key and next event key, `new_epoch` will
be `HummockEpoch::MAX`.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.
<!-- To manually trigger a benchmark, please check out [Notion](https://www.notion.so/risingwave-labs/Manually-trigger-nexmark-performance-dashboard-test-b784f1eae1cf48889b2645d020b6b7d3). -->

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<!-- 

You can ignore or delete the section below if you ticked the checkbox above.

Otherwise, remove the checkbox above and write a release note below.

-->

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

<!--
Please create a release note for your changes. 

Discuss technical details in the "What's changed" section, and 
focus on the impact on users in the release note.

You should also mention the environment or conditions where the impact may occur.
-->

</details>

Little-Wallace commented on src/storage/src/hummock/iterator/forward_user.rs:
```diff
@@ -103,7 +104,12 @@ impl<I: HummockIterator<Direction = Forward>> UserIterator<I> {
+                        {
+                            self.delete_range_iter.next();
+                        }
+                        if self.delete_range_iter.current_epoch() >= epoch {
```
would you mind moving this code to one method of `delete_range_iter `?

soundOfDestiny replied:
> would you mind moving this code to one method of `delete_range_iter `?

done

hzxa21 reviewed (commented):
The readability and maintainability is not good. I lose track while reviewing this PR so I think I should better stop leaving detailed comments on specific places and instead point out some general issues first:

- There is no meanful PR description. We can hardly know the motivation of this optimization and the general idea on what the optimization is before looking at the PR. What’s worse, even though we dig into the codes, it is not obvious to spot the inefficiency of the previous implementaion and what the new implementaion is. Some of us have participated in the slack discussion and can understand the motivation but this is not friendly to other reviewers, futher code development and the open source community. Unless the PR is related to minor refactoring or urgent hot fix, I suggest we spend some time summarizing the key motivation and changes in the PR description. This is useful not only for code reviews but also for tracking since some important key points may get forgotten and can hardly be found without a good PR/issue description.
- The structure of the newly added codes is not obvious and documentations are rare in this PR. Let’s take the newly added `build_events` method in `delete_range_aggregators.rs` as an exmple.
    - At the first glance, I can hardly understand what this method is used for since there is no documentation.
    - I try looking at the implementation. I know it is spliting the range tomestones by start key and end key as well as constructing some kind of reverse index. But still, I don’t know why we are doing this. For exmaple, I don’t know why `(start_user_key, 1, index)` is put in a vec called `delete` and why `(end_user_key, 0, index)` is put in a vec called `insert`.
    - I try to check the return type of `build_events` and look at what `CompactionDeleteRangeEvent` is. I only know it contains a user key + two vecs. But still, I don't know what this event stores since there is no documentation.
    - I try to look for the caller (`Sstable::new`) of this method to understand more. I saw that the returning tuple of `build_events` is used to construct a newly added member in Sstable called `monotonic_deletes`. I don't know what this is since the naming is not obvious and there is no documentation.
    - I then try to look at another place (`apply_event`) the returning tuple of `build_events` is used. I know it is doing some kind of dedup and vaugely understand what the two vecs are used for. Up till now, I already lose track and forget about why I am jumping around looking at all these codes so I get even more confused even though I finally find something that I can understand.
    
    This is just one example and I am already concern that if there is any bug in this PR, it can be hardly spotted. Even if this PR is 100% correct, it can be hardly maintained and potentially modified if there are future needs.
    
    I believe other reviewers will go through similar process and lose patience if they take a look at this PR. As a consequence, this PR will either be merged without taking a closer look or it will take a very long time to review. None of them are ideal. Therefore, I suggest we pay a little bit more attention to the code structure, naming and documentation, in addition to algorithm design and writing bug-free codes. After drafting the initial implementation, it is better to self-review the codes as a whole to see whether there is a lack of documentation, whether the structure makes sense and whether there are obvious bugs assuming we are the reviewer instead of the author. We can discuss together with other folks if we cannot come up with a good way to structure the codes. It is still more efficient than figuring it out in rounds of code reviews and revises.
    
- There is a unit test removed in this PR and no new unit test is added. Also, there is no e2e test or micro benchmark showing the impact. Testing and benchmarking is important for an optimization PR. Sometimes benchmarking can be avoided if the improvement is very obvious by looking at the implementation but this is not the case for this PR based on the points above.

soundOfDestiny commented on src/storage/src/hummock/compactor/mod.rs:
```diff
@@ -633,6 +637,19 @@ impl Compactor {
+            } else if earliest_delete_epoch < user_key_last_delete {
+                user_key_last_delete = earliest_delete_epoch;
+
+                iter_key.epoch = earliest_delete_epoch;
```
> IIUC, there is an invariant here: `iter_key.epoch < earliest_delete_epoch < user_key_last_delete` but this is not obvious before I diving deep into the implementation. Can we make this more explicit by adding an debug assert?
> 
> Also we should have better and consistent variable namings. How about `earliest_delete_epoch` -> `user_key_earliest_range_delete_epoch`, `user_key_last_delete` -> `user_key_last_delete_epoch`? Open to suggestions.

yes, you're correct.

soundOfDestiny replied:
> IIUC, there is an invariant here: `iter_key.epoch < earliest_delete_epoch < user_key_last_delete` but this is not obvious before I diving deep into the implementation. Can we make this more explicit by adding an debug assert?
> 
> Also we should have better and consistent variable namings. How about `earliest_delete_epoch` -> `user_key_earliest_range_delete_epoch`, `user_key_last_delete` -> `user_key_last_delete_epoch`? Open to suggestions.

fixed

soundOfDestiny commented on src/storage/src/hummock/compactor/mod.rs:
```diff
@@ -633,6 +637,19 @@ impl Compactor {
                 continue;
             }
 
+            if matches!(value, HummockValue::Delete) {
```
> nits: use `value.is_delete()` instead

fixed

soundOfDestiny commented on src/storage/src/hummock/compactor/mod.rs:
```diff
@@ -633,6 +637,19 @@ impl Compactor {
+
+                iter_key.epoch = earliest_delete_epoch;
+                sst_builder
+                    .add_full_key(iter_key, HummockValue::Delete, is_new_user_key)
```
> Please add documentation to explain why we add a Delete entry here for the same user key. I believe folks who didn't participate in the offline discussion can hardly understand it.

fixed

soundOfDestiny commented on src/storage/src/hummock/compactor/mod.rs:
```diff
@@ -594,6 +596,8 @@ impl Compactor {
                 last_table_id = Some(last_key.user_key.table_id.table_id);
             }
 
+            let earliest_delete_epoch = del_iter.earliest_delete(&iter_key.user_key, epoch);
```
> This is not the "earliest" delete epoch for the user key but the smallest epoch in range tombstones larger than the epoch of the current entry. Can you add more documentation to the method to explain it?

I have changed the variable and method name. Now the fact is obvious.

soundOfDestiny commented on src/storage/src/hummock/sstable/delete_range_aggregator.rs:
```diff
@@ -60,6 +61,30 @@ pub struct DeleteRangeAggregatorBuilder {
+type CompactionDeleteRangeEvent = (
+    UserKey<Vec<u8>>,
+    Vec<(usize, HummockEpoch)>,
+    Vec<(usize, HummockEpoch)>,
```
> What does the usize mean here? I think it is better to use a struct with named members instead of a type alias for a tuple here.

fixed

soundOfDestiny commented:
benchmark nexmark q14 result:

main:
```
Benchmark Name: nexmark-q14-blackhole-medium-1cn
Metrics Name: avg-source-output-rows-per-second
Result: NEGATIVE
Fluctuation: -29.500724% 
Result Value: 689460.438194
Baseline Value: 977968.109360
Execution ID: 2259
Baseline Execution ID: 2179
```

PR(4778362):
```
Benchmark Name: nexmark-q14-blackhole-medium-1cn
Metrics Name: avg-source-output-rows-per-second
Result: NEGATIVE
Fluctuation: -26.143716% 
Result Value: 722290.906182
Baseline Value: 977968.109360
Execution ID: 2257
Baseline Execution ID: 2179
```

soundOfDestiny commented on src/storage/src/hummock/sstable/mod.rs:
```diff
@@ -311,6 +362,10 @@ impl SstableMeta {
         }
+        buf.put_u32_le(self.monotonic_tombstone_events.len() as u32);
+        for monotonic_tombstone_event in &self.monotonic_tombstone_events {
+            monotonic_tombstone_event.encode(buf);
```
> Why do we need to encode `monotonic_tombstone_event`? It can be reconstructed by `range_tombstone_list`.

This is introduced in the (almost) latest commit 1324231b . We can revert the commit.

soundOfDestiny commented on src/storage/src/hummock/sstable/mod.rs:
```diff
@@ -271,6 +314,12 @@ pub struct SstableMeta {
+    /// can be transformed into events below:
+    /// `{ <0, +epoch1> <wmk1, -epoch1> <wmk1, +epoch2> <wmk2, -epoch2> <wmk2, +epoch3> <wmk3,
+    /// -epoch3> }`
+    pub monotonic_tombstone_events: Vec<MonotonicDeleteEvent>,
```
> Also it is better to explain what is monotonic. It seems that the event is ordered by (user key, epoch).

monotonic means ordered

soundOfDestiny replied:
> The documentation above is inaccurate since `MonotonicDeleteEvent` only contains a user key and epoch.

fixed

soundOfDestiny replied:
> Also it is better to explain what is monotonic. It seems that the event is ordered by (user key, epoch).

fixed

soundOfDestiny commented on src/storage/src/hummock/sstable/delete_range_aggregator.rs:
```diff
@@ -71,344 +104,303 @@ impl DeleteRangeAggregatorBuilder {
+                    1 => {
+                        enter.push(TombstoneEnterExitEvent {
+                            original_position_in_delete_tombstones: index,
+                            tombstone_epoch: delete_tombstones[index].sequence,
```
> Do we assume delete_tombstones is ordered by epoch?

of course not

soundOfDestiny commented on src/storage/src/hummock/sstable/delete_range_aggregator.rs:
```diff
@@ -71,344 +104,303 @@ impl DeleteRangeAggregatorBuilder {
-            let mut ret = tombstone.clone();
-            if !smallest_user_key.is_empty() && smallest_user_key.gt(&ret.start_user_key.as_ref()) {
-                ret.start_user_key = smallest_user_key.to_vec();
+        let (events, _) = DeleteRangeAggregatorBuilder::build_events(&tombstones_within_watermark);
```
> Why do we need special handling for `tombstones_within_watermark`? In other words, why not just return a combination of tombstones_within_watermark and tombstones_above_watermark?

We need to do gc or there will be more and more tombstones.

soundOfDestiny commented:
> The readability improves in the latest commit but it is still not easy to understand. I can get the basic idea after an offline chat with @Little-Wallace. If we cannot further simplify the codes, besides the comments I left, I think the thing we are missing in the code documentation are:
> 
> **1. The motivation of the changes** IIUC, it is to avoid storing excessive range tombstones after compaction if there are overlaps among range tombstones among different SSTs/batchs in compaction.
> 
> **2. The core idea of the changes** IIUC, the core idea contains two parts: 1) we only need to keep the smallest epoch of the overlapping range tomstone intervals since the key covered by the range tombstone in lower level must have smaller epoches; 2) due to 1, we lose the information to delete a key by tombstone in a single SST so we add a tombstone key in the data block.
> 
> **3. Examples or step-by-step documentations on the following topics:**
> 
> * The read process with range tombstone in the new design. In other words, how `ForwardMergeRangeIterator` works with `UserIterator`.
> * The compaction process with range tombstone in the new design. In other words, how `CompactionDeleteRangeIterator` works with compaction.
> * The build process of range tombstone in the new design. In other words, how to consolidate range tomestones in sst builder. What information are kept and persisted and what information are lost.

Fixed.
btw I have removed `event_seek_mapping` so the logic looks much more clear.

soundOfDestiny commented on src/storage/src/hummock/sstable/delete_range_aggregator.rs:
```diff
@@ -71,344 +104,303 @@ impl DeleteRangeAggregatorBuilder {
         self.delete_tombstones.extend(data);
     }
 
```
> Shall we rename `DeleteRangeAggregatorBuilder` to `CompactionDeleteRangesBuidler`?

fixed

soundOfDestiny commented on src/storage/src/hummock/sstable/mod.rs:
```diff
@@ -124,6 +126,46 @@ impl DeleteRangeTombstone {
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct MonotonicDeleteEvent {
+    event_key: UserKey<Vec<u8>>,
+    new_epoch: HummockEpoch,
```
> It will help a lot if you can add some documentation about what `new_epoch` means and show an example on how it will be used.

fixed

soundOfDestiny commented on src/storage/src/hummock/iterator/delete_range_iterator.rs:
```diff
@@ -203,49 +199,85 @@ impl ForwardMergeRangeIterator {
+        }
+        for node in &self.tmp_buffer {
+            let epoch = node.current_epoch();
+            if epoch != HummockEpoch::MAX {
```
> IIUC, we are using `HummockEpoch::MAX` to indicate a non-existent range tomestaone between previous `next_user_key()` and current `next_user_key()`, which is counter-intuitive and requires deep understanding in the implementation before figuring it out. Can we make it more explicit by making `current_epoch` return `Option<HummockEpoch>`?

* `Option` will bring extra cost since `HummockEpoch` is not a non-zero type.
* The default value of min delete range is `MAX`. Isn't it intuitive?

soundOfDestiny replied:
> Also, it is better to explicitly document `current_epoch()` indicates the range tomestone epoch between the previous `next_user_key()` and current `next_user_key()` to aid developers to understand the codes.

Fixed. I also added some explanation for the meaning of `HummockEpoch::MAX`.

soundOfDestiny commented on src/storage/src/hummock/sstable/mod.rs:
```diff
@@ -124,6 +126,46 @@ impl DeleteRangeTombstone {
+        apply_event(&mut epochs, &event);
+        monotonic_tombstone_events.push(MonotonicDeleteEvent {
+            event_key: event.0,
+            new_epoch: epochs.first().map_or(HummockEpoch::MAX, |epoch| *epoch),
```
> ditto: using `HummockEpoch::MAX` to indicate absence of a range tomestone seems counter-intuitive

`epochs.first()` means `min { epochs }`, so a default value `MAX` is intuitive.
I have added some explanation here.

Little-Wallace reviewed (approved):
LGTM
